### ISSUE TITLE
Could not load file or assembly System.Net.Security Version=4.0.0.0

### ISSUE THREAD

**Issue by spboyer** [StackExchange/StackExchange.Redis#350]
- Running coreclr rc1-update1
- webapi inside a docker container 
- throws exception on `ConnectionMultiplexer redis = ConnectionMultiplexer.Connect("redis:6379");`
- `redis` machine can be PING'ed from container
- running webapi outside of container can successfully read/write by IP 192.168.99.100:32783 (container IP:port)

`System.IO.FileNotFoundException: Could not load file or assembly 'System.Net.Security, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.`

**Comment by NickCraver** [comment]
How are you running the code, DNX? Relevant issues: https://github.com/dotnet/corefx/issues/4533 and https://github.com/aspnet/dnx/issues/3183
TL;DR (those are pretty long): There was a bug on rc1-final of DNX, it's only fixed in later rc2 builds.

We'll definitely be doing an RC2 update here as soon as it's nearing to fix many things missing or potentially broken in RC1, but we'll need the Microsoft bits in place before we can update on the lib side.

**Comment by spboyer** [comment]
Thanks @NickCraver - i'll hurry up and wait.  This was for demo / blog purpose only at this point.

**Comment by spboyer** [comment]
@nickcraver the 2 issues referenced are now closed, can we get a new release?

cc:/ @benadams

**Comment by NickCraver** [comment]
@spboyer new release is up on NuGet: 1.1.572-alpha (we'll being dogfooding this today): https://www.nuget.org/packages/StackExchange.Redis/1.1.572-alpha

**Comment by spboyer** [comment]
just gave it a run
```
 Microsoft.AspNet.Server.Kestrel[13]
      An unhandled exception was thrown by the application.
      System.IO.FileNotFoundException: Could not load file or assembly 'System.Net.Security, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
      
      File name: 'System.Net.Security, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' ---> System.IO.FileNotFoundException: Could not load the specified file.
      File name: 'System.Net.Security'
         at System.Runtime.Loader.AssemblyLoadContext.LoadFromAssemblyName(AssemblyName assemblyName)
         at System.Runtime.Loader.AssemblyLoadContext.Resolve(IntPtr gchManagedAssemblyLoadContext, AssemblyName assemblyName)
         at StackExchange.Redis.ConfigurationOptions.Clear()
         at StackExchange.Redis.ConfigurationOptions.DoParse(String configuration, Boolean ignoreUnknown)
         at StackExchange.Redis.ConnectionMultiplexer.CreateMultiplexer(Object configuration)
         at StackExchange.Redis.ConnectionMultiplexer.ConnectImpl(Func`1 multiplexerFactory, TextWriter log)
         at webapi.Controllers.HeroesController..ctor()
         at lambda_method(Closure , IServiceProvider , Object[] )
         at Microsoft.AspNet.Mvc.Infrastructure.DefaultTypeActivatorCache.CreateInstance[TInstance](IServiceProvider serviceProvider, Type implementationType)
         at Microsoft.AspNet.Mvc.Controllers.DefaultControllerFactory.CreateController(ActionContext actionContext)
         at Microsoft.AspNet.Mvc.Controllers.FilterActionInvoker.<InvokeAllActionFiltersAsync>d__52.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Microsoft.AspNet.Mvc.Controllers.FilterActionInvoker.<InvokeExceptionFilterAsync>d__51.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at Microsoft.AspNet.Mvc.Controllers.FilterActionInvoker.<InvokeAsync>d__44.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Microsoft.AspNet.Mvc.Infrastructure.MvcRouteHandler.<RouteAsync>d__6.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at Microsoft.AspNet.Mvc.Routing.InnerAttributeRoute.<RouteAsync>d__10.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at Microsoft.AspNet.Routing.RouteCollection.<RouteAsync>d__9.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Microsoft.AspNet.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Microsoft.AspNet.IISPlatformHandler.IISPlatformHandlerMiddleware.<Invoke>d__8.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Microsoft.AspNet.Hosting.Internal.RequestServicesContainerMiddleware.<Invoke>d__3.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
         at Microsoft.AspNet.Hosting.Internal.HostingEngine.<>c__DisplayClass32_0.<<Start>b__0>d.MoveNext()
      --- End of stack trace from previous location where exception was thrown ---
         at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
         at Microsoft.AspNet.Server.Kestrel.Http.Frame.<RequestProcessingAsync>d__79.MoveNext()
```

**Comment by spboyer** [comment]
@NickCraver Here is the use case.  When running under DNX on local OSX it is a non issue.  The Security exception happens when running the application within a Docker Container.  Here is the repo where you can see the code reduces to the `/api/values` -> https://github.com/spboyer/redis-api

**Comment by spboyer** [comment]
@markrendle @glennc ideas here guys?

**Comment by jeremymeng** [comment]
@spboyer what's the output of `dnx --version` from the docker image?

**Comment by spboyer** [comment]
@jeremymeng running this base image
microsoft/aspnet:1.0.0-rc1-update1-coreclr

**Comment by nishankbangera** [comment]
Even I am getting same error when running from docker. also tried with ubuntu os instead of debian still same error.

dnx version - 1.0.0-rc1-16202 
type - coreclr

**Comment by jeremymeng** [comment]
I have seen a similar exception before on Debian, and was able to workaround by modifying /etc/os-release to have ubuntu ID's. Note that the ordering matters.

PRETTY_NAME="Debian GNU/Linux 8 (jessie)"
NAME="Debian GNU/Linux"
VERSION="8 (jessie)"
**ID=ubuntu
VERSION_ID="14.04"**
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

**Comment by jeremymeng** [comment]
Even the `microsoft/aspnet:1.0.0-rc1-update1-coreclr` docker image sets the `DNX_RUNTIME_ID` environment variable, it seems that `dnu restore` was not able to restore a couple of runtime nuget packages on Debian. 

I just tested that the following workaround works

1. add a file `os-release.ubt` with content mentioned above
2. add a line in `Dockerfile` above the dnu restore line
```
COPY os-release.ubt /usr/lib/os-release
```

**Comment by nishankbangera** [comment]
After making above changes. dnu restore failed for most of the nuget packages

![untitled](https://cloud.githubusercontent.com/assets/16240705/13175364/fcde21ce-d72f-11e5-8268-9c0400d98178.png)

**Comment by jeremymeng** [comment]
@nishankbangera my workaround is specific to Debian.  If you also see the exception on ubuntu it might be a different issue.

**Comment by nishankbangera** [comment]
@jeremymeng I had tried it with debian OS. base image used was microsoft/aspnet:1.0.0-rc1-final-coreclr.  I had copied  file "os-release.ubt" to " /usr/lib/os-release",  do I also need to place it in "/etc/os-release" as stated in one of your previous post

**Comment by jeremymeng** [comment]
@nishankbangera No you don't need to.  In that base image `/etc/os-release` is a symbolic link to `/usr/lib/os-release`.

**Comment by raffaeler** [comment]
I still have this error and did not understand which is the best workaround....

If I reference: "System.Net.Security": "4.0.0-beta-23516",
The result is:
System.IO.FileNotFoundException: Could not load file or assembly 'System.Net.Security

If instead I reference: "System.Net.Security": "4.0.0-rc2-23826",  (taken from the aspnetcidev feed)
dnu fails restoring the package during Docker deploy with the following error:
"Unable to restore packages"

Any hints please?

**Comment by rudygt** [comment]
I ran into this problem using the same docker image https://github.com/aspnet/aspnet-docker/blob/master/1.0.0-rc1-update1-coreclr/Dockerfile 

I was able to get it to work adding this to the project.json dependencies 

 "System.Net.Security": "4.0.0-beta-*",
 "Microsoft.NETCore.Platforms": "1.0.1-beta-*"

and using

dnu restore --runtime=ubuntu.14.04-x64

**Comment by raffaeler** [comment]
good to know, thank you. I'll give a try

**Comment by wisien92** [comment]
I aslo need System.Net.Security - how did you pass dnu restore with flags in dockerimage? 

This is my dockerfile:

FROM microsoft/aspnet:1.0.0-rc1-update1-coreclr
COPY . /app
WORKDIR /app
RUN ["dnu", "restore --runtime=ubuntu.14.04-x64"]
COPY . /app
EXPOSE 5004
ENTRYPOINT ["dnx", "-p", "project.json", "web"]

**Comment by raffaeler** [comment]
forget about RC1 and go with RC2
DNU is dead, long live dotnet CLI

**Comment by wisien92** [comment]
So how should my dockerfile look? I have been following onbuild https://github.com/dotnet/dotnet-docker/tree/master/1.0.0-rc2 and I can run my docker but I cannot view page.

**Comment by glennc** [comment]
ASP.NET Core doesn't allow connections from anything other than localhost by default. If you add `UseUrls("http://*:5004")` or something like that to your program.cs then you should be good.

**Comment by wisien92** [comment]
@glennc That's perfect! To make it easier I came across these docker tools for VS and RC2 https://visualstudiogallery.msdn.microsoft.com/0f5b2caa-ea00-41c8-b8a2-058c7da0b3e4 (maybe someone will find them usefull)

And sory for offtopic but I have one last problem - SignalR Hub Context. I have added
`services.AddSignalR();` and `app.UseWebSockets(); app.UseSignalR();` but in my main controller in method

```
private IHubContext loadingHub;
private IHubContext centralHub;

public HomeController(IConnectionManager connectionManager, IHostingEnvironment appEnvironment)
{
    loadingHub = connectionManager.GetHubContext<LoadingHub>();
    centralHub = connectionManager.GetHubContext<CentralViewHub>();
}
```

both loadingHub and centralHub are null - and I need a way to send message to all or single client from controller - am I missing some kind of mapping that wasnt required?

**Comment by NickCraver** [comment]
There's so much .NET Core churn we're trying to keep up with - is this still an issue in the latest version?

**Comment by raffaeler** [comment]
Not at all. If you use the RTM releases, you should never have this problem.
Also the library is much bigger now (netstandard1.6), and should be doubled (in terms of # of api) with the upcoming netstandard2.0.

**Comment by NickCraver** [comment]
@raffaeler thanks for the info. Given the latest releases are on the RTM packages, I'll close this out. If there's still an issue we're unaware of - please ping and I'll reopen. I'm trying to get issues in order on a number of libraries for the .NET 4.0 drop, netstandard2.0 prep, and .csproj fun.
